cron/jobstats.go

```diff
@@ -0,0 +1,117 @@
+package cron
+
+import (
+	"database/sql"
+	"fmt"
+)
+
+type Job struct {
+	Category	string
+	Status		string
+	Count		int
+}
+
+//jobs/submitted
+func GetSubmittedJobCounts(db *sql.DB, startDay string, endDay string)([]Job, error) {
+	var jobType *string
+	var count int
+	//skip curl wrapper jobs
+	query := `SELECT b.job_type, count(b.*) AS count
+            FROM (
+              SELECT a.id,
+                CASE WHEN array_length(a.types, 1) = 1 THEN a.types[1]
+              ELSE 'Mixed'
+           END AS job_type
+           FROM (
+               SELECT j.id, array_agg(DISTINCT t.name) AS types FROM jobs j
+               JOIN job_steps s ON j.id = s.job_id
+               JOIN job_types t ON s.job_type_id = t.id
+               WHERE j.start_date >= ($1 :: DATE) AND j.start_date <= ($2 :: DATE)
+               AND j.app_id != '1e8f719b-0452-4d39-a2f3-8714793ee3e6'
+               GROUP BY j.id
+           ) AS a
+          ) AS b
+         GROUP BY b.job_type
+         ORDER BY b.job_type;`
+
+	rows, err := db.Query(query, startDay, endDay)
+	if err != nil {
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	var jobs []Job
+
+	for i := 0; rows.Next(); i++ {
+		err := rows.Scan(&jobType, &count)
+		if err != nil {
+			return nil, err
+		}
+		jobs = append(jobs, Job{getStringValue(jobType), "Submitted", count})
```

Comment:
Are you sure this works right? Given that `count` and `jobType` are initialized outside the loop, if they're not being passed by value when creating the `Job`, you could end up with an array of a bunch of the same data (the data for the last row).

Also, if you initialize a `Job` at the start of the loop, you can probably scan straight into the struct instead of scanning into intermediate variables, by something like:

```
for rows.Next() {
    var job Job
    if err := rows.Scan(&job.Category, &job.Status, &job.Count); err != nil {
        ... as before ...
    }
    jobs = append(jobs, job)
    ... as before ...
}
```

You'd probably also want to change the DB code to pull the "Submitted" from the DB. You can put static strings into column lists for `SELECT`, which would probably be the best way to start on this for now (which also opens things up for easier extensibility to other statuses than submitted later, probably).

Reply:
It does work but I see what you're saying, there are a lot of small opportunities to be more efficient, I like what you proposed so I will change it to that.